### ISSUE TITLE
I failed to get the prediction results in the paper

### ISSUE THREAD

**Issue by 705062791** [FraLuca/STSGCN#12]
I failed to get the prediction results in the paper, while used the proposeed code and pretrained model. I speculate that a possible reason is: **Using my own test code**, due to its absence in the public code.

**My test method is as follows**: 
Calculate the average 3D joints error in each action per time step as shown in paper. For instance, In 'walking’, I calculate the mean 3D error at 40ms, 160ms, 320ms and 400ms. The perdiction errors are significantly worse than the results published in the paper.

To address above issue, I think the best way is to publish your test code in the paper.

**Comment by FraLuca** [comment]
Hi,

Thanks for your interest!
In the readme markdown file, there are all the instructions to run the test easily.
In particular an example could be:

`python main_h36_3d.py --input_n 10 --output_n 25 --skip_rate 1 --joints_to_consider 22 --mode test --model_path ./checkpoints/CKPT_3D_H36M`

I want to stress `--mode test` setting that switches the model to test mode.
Let me know if you can solve it.

Luca

**Comment by 705062791** [comment]
Hi
Thank you for your reply.
I did notice the content in the README, and I also used the command you mentioned to run the code. 

However, the test part only prints the average 3D error of each action, but I expect to get the **frame by frame 3D error** as given in your paper **table1**.

So, I modified the test function to calculate the error frame by frame, the result is:
40ms      160ms    320ms    400ms
17.7350  33.9029  56.3370, 67.4853

And in your paper, it is:
40ms      160ms    320ms    400ms
10.1        17.1        33.1       38.3

I doubt if my test code is wrong, so I hope you can provide the test code to calculate the frame-by-frame error instead of the action mean error.

Looking forward to your reply!

Tiezheng Ma

**Comment by FraLuca** [comment]
In Table 1 (short-term) and Table 2 (long-term) we show the MPJPE for each action.
This metric considers the error in all the previous frames w.r.t. a considered one and averaging them. For example, if we consider the prediction to be 80 msec (2 frames) we run the error for frame 1 and frame 2 and then we average them. This gives the results in the tables you are referring to.
Moreover, we adopt this test function/metric coherently with the standard procedure of the literature (in order to be comparable with previous works). In particular, you can refer to the baseline model of Wei Mao [History Repeat Itself](https://arxiv.org/pdf/2007.11755.pdf).

Luca

**Comment by research-lover** [comment]
Hi,

I think previous methods evaluate on the specific frames not the average over all previous frames. 

It is not fair, if you compare your results to them. Please refer the evaluation code of [Residual.sup](https://github.com/una-dinosauria/human-motion-prediction) and [Learning Trajectory Dependencies](https://github.com/wei-mao-2019/LearnTrajDep).

Thanks

Research Lover

**Comment by theodoriss** [comment]
Hi,

Please check the formulas 5 and 6 in  [LearnTrajDep](https://arxiv.org/pdf/1908.05436.pdf) .
The error is averaged over the number of  frames.

Theodoros

**Comment by FraLuca** [comment]
Hi Research Lover,

Thank you for your point. We'll double-check and, in case, take action.
We'll let you know soon.

Thanks, 
Luca

**Comment by Dean-UQ** [comment]
Hi, 
It is true that this work did not follow the common evaluation setup used in other related works. Hence, it is an unfair to compare its performance with other SOTA methods.

**Comment by FraLuca** [comment]
Hi,
We found the problem in the evaluation metric.
We will provide the correct tables and results soon, here and in the paper.
Thanks for your help.

**Comment by Maradowei** [comment]
Hi,
I don't understand the eq::6 and eq::7, why the k is from 1 to T+K, rather than T to T+K, and the denominator same. 
I mean why the input are under the considering for Loss.
Good work/paper, by the way.

**Comment by Dean-UQ** [comment]
【自动回复】同学你好，课程作业已收到！

**Comment by whf9527** [comment]
How should I get AMASS data from this website
![image](https://user-images.githubusercontent.com/43332051/167237918-a981a8ba-1ed3-4e6c-ae97-6087d7b1acd1.png)

**Comment by znqaq** [comment]
> 

Using the code with output=25 to train and then calculate the loss frame by frame, I got similar results to yours, but his code uses 32 nodes in the test set, but the test set is 22 nodes, is this a problem?

**Comment by dulucas** [comment]
Hi,

We've been waiting you to take action but until now nothing happened..The paper is still unchanged.

Do you plan to withdraw your paper or make an official annocement? There are already some recent papers used your code and make unfair comparison with previous works, which is clearly not good for the community.

Best

**Comment by FraLuca** [comment]
Sorry for only replying now.

The problem arises because no prior human pose forecasting work has explicitly written the test MPJPE metric. [Mao et al., 2020, Mao et al., 2019] have specified the MPJPE for the learning loss, and they have referred to the (same) MPJPE for testing, which is however different.

In [Mao et al., 2020], Eq. (6), they define MPJPE as 

$$MPJPE = \frac{1}{J(M+T)}\sum_{t=1}^{M+T} \sum_{j=1}^J ||\hat{\textbf{p}}_{t,j} - \textbf{p}_{t,j} ||^2,$$

which sums up all errors at all frames up to the prediction T.

Also in [Ionescu et al., 2014], Eq. (8), they define the MPJPE as:

$$MPJPE(t) = \frac{1}{J} \sum_{j=1}^J ||\hat{\textbf{p}_{t,j} }- \textbf{p}_{t,j} ||^2,$$

and they state: "For a set of frames the error is the average over the MPJPEs of all frames."

We have therefore interpreted the test MPJPE to be:

$$MPJPE = \frac{1}{J T}\sum_{t=M+1}^{M+T} \sum_{j=1}^J ||\hat{\textbf{p}}_{t,j} - \textbf{p}_{t,j} ||^2,$$

which is implemented in our testing code. Note: coding has been done in good faith, and in good faith we have open-sourced the project here.

As noted in this thread, the code provided by [Mao et al., 2020] actually considers only the target temporal horizon, not the average up to that time.

Running the test code of [Mao et al., 2020], short-term (400ms) and long-term (1000ms) errors for the Human3.6M dataset for STS-GCN are:\\
TABLE with STS-GCN  (maybe just our method, not all)

The main page in this repository will report this performance and specify the test MPJPE error, to avoid future discrepancies.

![image](https://user-images.githubusercontent.com/44803179/177993948-c3a67077-7369-4758-a169-f912e0d70e97.png)

**Comment by dulucas** [comment]
OK, but why you deleted my comments?

**Comment by dulucas** [comment]
And by the way, I saw you used some codes from LTD's repo(like the dataloader and utils etc), but why you only didn't use the testing code? I have saved this page in PDF in case you delete my comments again..

**Comment by znqaq** [comment]
老哥这波重捶nb，我也有同样的疑惑，这个作者写了很好的代码框架但测试代码缺不一样，我前段时间请教了那位Mao Wei，他认为目前该领域都是预测特定帧，而不是之前帧的平均，平均之后会使得后面帧的损失影响降低导致不公平的比较。
&nbsp;


zn
***@***.***



&nbsp;




------------------&nbsp;原始邮件&nbsp;------------------
发件人: "Yuming ***@***.***&gt;; 
发送时间: 2022年7月8日(星期五) 晚上9:05
收件人: ***@***.***&gt;; 
抄送: ***@***.***&gt;; ***@***.***&gt;; 
主题: Re: [FraLuca/STSGCN] I failed to get the prediction results in the paper (Issue #12)





 
And by the way, I saw you used some codes from LTD's repo(like the dataloader and utils etc), but why you only didn't use the testing code? I have saved this page in PDF in case you delete my comments again..

**Comment by Dean-UQ** [comment]
This paper will indeed mislead the future development of the human motion prediction task or community clearly. I strongly suggest that the author revise their official paper published on ICCV. If not, I will send an e-mail to the ICCV program chair to state the intolerable mistake in this paper.

**Comment by ReasearchJudge** [comment]
For those who are studying motion prediction and might be confusing about some papers with "awsome results",
be careful that there are some works of motion prediction which are using unfair testing protocols:
- testing on "average result of multiple frames" but compare with results of other methods which are tested on one certain frame:
[1] Space-Time-Separable Graph Convolutional Network for Pose Forecasting, iccv2021 (this repo)
[2] Spatio-Temporal Gating-Adjacency GCN for Human Motion Prediction, CVPR2022

- tested on 17 joints but compare with results of other works which are tested on 22 joints:
[3] Learning Dynamic Relationships for 3D Human Motion Prediction, cvpr202
[4] Motion prediction using trajectory cues, iccv2021 (this work is even more "interesting": this works train 4 different models (for x, y, z, v separately) for "each action", so they have 60 models for Human3.6m dataset, but they compare their result with other methods using a single model for all the actions)

List to be (but hope not) updated...

**Comment by Dean-UQ** [comment]
Dear readers of STSGCN,
It is clear that there are some mistakes represented in this paper. Since the author did not declare and revise their published paper until now, it is necessary to reflect this issue to the program chair of ICCV-21. So, please leave your comments her[e](https://github.com/FraLuca/STSGCN/issues/18l) to make the chair know the impartial peer reviews from the human motion prediction community.
Thanks for your contributions to the community.

**Comment by FraLuca** [comment]
Let us close this GitHub issue because the previous post clarified the discrepancy between the two versions of the error metric.
We wish to thank those who pointed out the issue: it is always a good thing when we can clarify a mistake.
We are happy that having open-sourced the project has enabled it.
